File: 3619_Count_Islands_With_Total_Value_Divisible_by_K.py
class Solution(object):
    def countIslands(self, grid, k):
        """
        :type grid: List[List[int]]
        :type k: int
        :rtype: int
        """
        
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]

        m, n = len(grid), len(grid[0])

        ans = 0;

        def dfs(i, j, m, n, grid):
            if i < 0 or j < 0 or i == m or j == n:
                return 0
            if grid[i][j] == 0:
                return 0

            val = grid[i][j] % k
            grid[i][j] = 0

            for k in range(4):
                val = (val + dfs(i + dx[k], j + dy[k], m, n, grid)) % k

            return val

        for i in range(m):
            for j in range(n):
                if grid[i][j] > 0:
                    tot = dfs(i, j, m, n, grid)
                    if tot % 2 == 0:
                        ans += 1

        return ans

File: cpp/3619_Count_Islands_With_Total_Value_Divisible_by_K.cpp
class Solution {
public:
    int dx[4] = {-1, 1, 0, 0};
    int dy[4] = {0, 0, -1, 1};

    int dfs(int i, int j, int m, int n, vector<vector<int>>& grid, int k)
    {
        if (i < 0 || j < 0 || i == m || j == n) return 0;
        if (grid[i][j] == 0) return 0;

        int val = grid[i][j] % k;
        grid[i][j] = 0;

        for (int d = 0; d < 4; ++d)
        {
            // (a + b) % k -> ((a % k) + (b % k)) % k
            val = (val + dfs(i + dx[d], j + dy[d], m, n, grid, k)) % k;
        }
        return val;
    }

    int countIslands(vector<vector<int>>& grid, int k) {
        int m = grid.size(), n = grid[0].size();
        int ans = 0;

        for (int i = 0; i < m; ++i)
        {
            for (int j = 0; j < n; ++j)
            {
                if (grid[i][j] > 0)
                {
                    int tot_mod = dfs(i, j, m, n, grid, k);
                    if (tot_mod == 0)
                        ++ans;
                }   
            }
        }
        return ans;
    }
};
